qa/vagrant/src/main/java/org/elasticsearch/packaging/util/ServerUtils.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.packaging.util;
+
+import org.apache.http.HttpResponse;
+import org.apache.http.client.fluent.Request;
+import org.apache.http.conn.HttpHostConnectException;
+import org.apache.http.entity.ContentType;
+import org.apache.http.util.EntityUtils;
+
+import java.io.IOException;
+import java.util.Objects;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsString;
+
+public class ServerUtils {
+
+    public static void waitForElasticsearch() {
+        waitForElasticsearch("green", null);
+    }
+
+    public static void waitForElasticsearch(String status, String index) {
+
+        Objects.requireNonNull(status);
+
+        // httpclient does not appear to support waiting between retries when the connection is refused, so we loop here
+        boolean started = false;
+        for (int i = 0; i < 120; i++) {
+            try {
+
+                final HttpResponse response = Request.Get("http://localhost:9200/_cluster/health")
+                    .connectTimeout(120000)
+                    .socketTimeout(120000)
+                    .execute()
+                    .returnResponse();
+
+                if (response.getStatusLine().getStatusCode() >= 300) {
+                    final String statusLine = response.getStatusLine().toString();
+                    final String body = EntityUtils.toString(response.getEntity());
+                    throw new RuntimeException("Connecting to elasticsearch cluster health API failed:\n" + statusLine+ "\n" + body);
+                }
+
+                started = true;
+                break;
+
+            } catch (HttpHostConnectException e) {
+                // we want to retry if the connection is refused, so do nothing here
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+
+            try {
+                Thread.sleep(1000);
```

Comment:
I'm no expert, but I buy the argument [here](https://stackoverflow.com/questions/3976344/handling-interruptedexception-in-java) that you should probably re-interrupt the thread just in case. The runtime exception is fine though because you want the test to fail.

Reply:
Honestly why not let the `InterruptedException` bubble out and let the test framework catch it and do whatever it thinks is right with it?

Reply:
I would argue against no sleep at all. If we can make the tests that much faster burning the CPU is worth it. Hammering the port during startup might even uncover something interesting. Would probably need to switch from to a time based loop instead of a number of retries based one.

Reply:
Yeah where I didn't have strong opinions I just followed what the bats tests did. 1 second intervals with 120 retries seems a little excessive
https://github.com/elastic/elasticsearch/blob/631a53a0e1d6be928f6c7bb55bc4c64de5bba695/qa/vagrant/src/test/resources/packaging/utils/utils.bash#L482

I would argue that we do want to set some lower time bound on how long we wait for the server to come up, because that's always going to take significantly longer than what we're doing here whether we retry or not. That is, while the server startup time varies, it's never going to occasionally be only a millisecond. Does that make sense or am I misunderstanding?

My thinking on InterruptedException was just that we should fail fast, and that I didn't want to add it to the signature of every method in the call stack. That's a good point about letting junit handle it - I took a (not very deep) look and it doesn't seem like junit 4 does anything interesting with it. I'll think a little more about what to do here but I'll at least reset the interrupt on the thread

Reply:
The server will take some time to come up for sure, but why try to guess how much that is, when the cost of not doing so is not significant.  If this were a communication in production, then we would have to do all these things to be nice to the server, but since this is a test, we don't need to be nice. All we need is to measure how long we have been trying to connect for and time out if it's too long, but we don't need necessarily need to add delays in between. 

I don't think not dealing with the interrupt is is not likely to interfere with junit, I never saw it do, but it's a bad practice in general, as the interrupt is never reset, it gets lost and the thread that was interrupted will keep on running. I found this to be a big deal when it's in some utility or library that the thread calls, and then one is left to wonder why graceful termination is not possible elsewhere

Reply:
Gotcha, I'll remove the sleep and terminate the waiting based on directly measuring the time spent